doc/designs/quic-design/synthesis.md

```diff
@@ -0,0 +1,489 @@
+__owur int SSL_get_next_tick(SSL *ssl);
+
+/*
+ * Give libssl an opportunity to do things, for example due to timer events,
+ * even when no specific read or write call is being performed.
+ *
+ * Note that:
+ *   - For TLS this ss a no-op, as there are no timer events for TLS;
+ *   - For DTLS this is equivalent to calling DTLSv1_handle_timeout();
+ *   - For QUIC, only SSL objects representing a connection have timer events;
+ *   - For QUIC SSL objects representing additional streams, this is a no-op.
+ *
+ * Returns 1 on success and 0 on failure.
+ */
+__owur int BIO_tick(BIO *bio); /* BIO_C_TICK */
+__owur int SSL_tick(SSL *ssl);
+
+/*
+ * Get the SSL object representing the connection associated with this object.
+ *
+ * If the SSL object represents a non-QUIC method or a QUIC connection, this
+ * returns the same object passed.
+ *
+ * If the SSL object represents a QUIC stream returns the QUIC connection
+ * object.
+ */
+__owur SSL *SSL_get0_connection(SSL *ssl);
+
+/*
+ * Returns 1 if the object represents a connection. This always returns 1 for
+ * non-QUIC methods, but returns 0 for SSL objects for QUIC streams which are
+ * not also the QUIC connection object.
+ *
+ * This is exactly equivalent to (SSL_get0_connection(ssl) == ssl).
+ */
+__owur int SSL_is_connection(SSL *ssl);
+
+/*
+ * If the object represents a stream, returns a SSL_STREAM_TYPE value
+ * designating whether the stream can be used for transmission, reception,
+ * or both.
+ *
+ * This always returns SSL_STREAM_TYPE_BIDI for non-QUIC methods.
+ *
+ * It returns SSL_STREAM_TYPE_NONE for a QUIC connection object if it
+ * does not have a default stream.
+ */
+#define SSL_STREAM_TYPE_NONE    0
+#define SSL_STREAM_TYPE_READ    1
+#define SSL_STREAM_TYPE_WRITE   2
+#define SSL_STREAM_TYPE_BIDI    (SSL_STREAM_TYPE_READ | SSL_STREAM_TYPE_WRITE)
+__owur int SSL_get_stream_type(SSL *ssl);
+
+/*
+ * QUIC: Returns the unique stream ID for the stream, an integer in range [0,
+ * 2**62-1]. If called on a QUIC connection, returns the unique stream ID for
+ * the default stream if there is one, and otherwise returns 0.
+ *
+ * TLS, DTLS: Returns 0.
+ * TODO: Previously Matt proposed this just returns a monotonically incrementing
+ *       counter for TLS/DTLS, but is there any reason to implement this?
+ */
+__owur uint64_t SSL_get_stream_id(SSL *ssl);
+
+/*
+ * Returns the SSL object's preference for a maximum UDP payload size.
+ *
+ * A client can use this to determine the size of the buffers they should use
+ * for receiving datagrams to be pushed to the SSL object. This is only
+ * necessary if the client is shunting its own buffers to libssl, and only if
+ * support for larger than worst case MTUs is desired. The returned value should
+ * be set as the MTU on the BIO_dgram to indicate to the SSL object that support
+ * for larger MTUs is available.
+ */
+__owur ssize_t SSL_get_preferred_max_dgram_len(SSL *ssl);
```

Comment:
I currently fail to see why we need this for an MVP.

Reply:
Why is this `ssize_t` ?

Reply:
I guess so that it can fail. But I guess it could return 0 on failure.